roc_test.go

```diff
@@ -0,0 +1,53 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import (
+	"math/rand"
+	"testing"
+)
+
+func TestROC(t *testing.T) {
+
+	cases := []struct {
+		y   []float64
+		pos []bool
+		tpr []float64
+		fpr []float64
+	}{
+		{
+			y:   []float64{0.0, 1.0, 2.0, 3.0, 4.5, 5.0, 5.5, 6.0, 7.0, 8.0, 9.0, 9.5},
+			pos: []bool{true, true, true, false, true, false, false, false, false, false, false, false},
+			tpr: []float64{0.0, 0.25, 0.5, 0.75, 0.75, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0},
+			fpr: []float64{0.0, 0.0, 0.0, 0.0, 0.125, 0.125, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 1.0},
+		},
+		{
+			y:   make([]float64, 12),
+			pos: make([]bool, 12),
+			tpr: []float64{0.0, 0.25, 0.5, 0.75, 0.75, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0},
+			fpr: []float64{0.0, 0.0, 0.0, 0.0, 0.125, 0.125, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 1.0},
+		},
+	}
+	p := rand.Perm(len(cases[0].y))
+	for i, j := range p {
+		cases[1].y[i] = cases[0].y[j]
+		cases[1].pos[i] = cases[0].pos[j]
+	}
+	for _, c := range cases {
+		tpr, fpr := ROC(c.y, c.pos)
+		if len(tpr) != len(fpr) {
+			t.Fatalf("len(tpr) (%d) does not equal len(fpr) (%d)", len(tpr), len(fpr))
+		}
+		if len(tpr) != len(c.tpr) {
+			t.Fatalf("len(tpr) (%d) does not equal expected length (%d)", len(tpr), len(c.tpr))
+		}
+		for i, _ := range tpr {
```

Comment:
Replace for loop (should omit `, _` BTW) with:

```
if !reflect.DeepEqual(gotTpr, c.wantTpr) {
    t.Errorf("unexpected TPR got:%v want:%v", gotTpr, test.wantTpr)
}
if !reflect.DeepEqual(gotTfr, c.wantTfr) {
    t.Errorf("unexpected TFR got:%v want:%v", gotTfr, test.wantTfr)
}
```

If you add `i` to the cases for loop, you can report which test failed here too.